### ISSUE TITLE
[Feature request]: Add support for configuring docker build args in Jenkins UI.

### ISSUE THREAD

**Issue by skamenetskiy** [jenkinsci/docker-plugin#730]
Add support for configuring docker build args in Jenkins UI.

**Comment by pjdarton** [comment]
Whilst I'm usually a big fan of brevity, in this case it'd be useful to have more detail - there's a lot of Jenkins UI, and which bit you mean will have a dramatic effect on the feasibility of doing this and/or whether or not it's really in-scope for this plugin (e.g. lots of folks don't [RTFM](https://github.com/jenkinsci/docker-plugin/blob/master/README.md) and go to the wrong place).

**Comment by GyrosGeier** [comment]
I just stumbled over this as well. We have a few Dockerfiles that can be parameterized through `ARG` arguments, and there is currently no way to specify values (the `--build-args` parameter to `docker build`).

**Comment by GyrosGeier** [comment]
FWIW, the same question has been asked in #556 .

**Comment by pjdarton** [comment]
Whether that was the same question or not is in some doubt - that issue was resolved without any changes to the docker-plugin.
I think #556 was actually raised in the wrong place - it was reporting a problem that came from within the docker daemon (not from the Jenkins docker-plugin), as the `ARG` declaration and its usage were both in the `Dockerfile` and the docker-plugin had successfully passed all that to the docker daemon for processing - the error came from the docker daemon rejecting the Dockerfile ... and later versions of docker were able to correctly process the Dockerfile.  The Jenkins plugin never needed to know about any `ARG` functionality in that case.

Given my last request for clarification went unanswered ... can you confirm that you're referring to the Jenkins docker-plugin "Build Step" that passes a Dockerfile (and files) to the docker daemon (the equivalent of `docker build`)?
If so then, sure, that's the sort of enhancement I'd agree to, and would welcome (and assist with) a pull request (aka "PR") that implemented it.  If you're up for that then I'd suggest you read https://github.com/jenkinsci/docker-plugin/blob/master/CONTRIBUTING.md

**Comment by GyrosGeier** [comment]
Yes, I am referring to that. I can put this on the list of things to look at, but I have little to no experience with Jenkins plugins, so it'd take me a few weeks to understand what is going on before I can make a patch.

**Comment by pjdarton** [comment]
FYI you can probably find some good examples in previous pull requests.
Key aspects include:
1) figure out what class is responsible for the functionality
2) add a new field to store the new data
3) add a getter and a databoundsetter setter for that field to cater for XML serialisation
4) add getXXXAsString & matching setter that provide/take the data as a (multi-line) string (as that'll make webui support easier)
5) find the config.jelly file for the class and add a new multi-line string field xxxAsString
6) create a html help file for the new field (put it alongside the jelly file)
7) add a validation method to the class' Descriptor class that'll get called every time the user edits that new field ... and make it tell the user where they went wrong!
8) edit the "do the docker build step" code and make it take account of the new field data (if present) ... ensuring that if the new field is null/empty then nothing new happens.
9) write unit tests for all the new stuff
10) try it out using `mvn hpi:run` and check it builds and tests with `mvn install`

There's also a gazillion other docs on the subject (some more up to date than others); I think GitHub is the primary source of truth now.
Ps. For an easy life, compile & run on Linux not Windows.

**Comment by Elbandi** [comment]
how to add the build arg at jenkins ui?

![image](https://user-images.githubusercontent.com/44674/145399793-e04d8c8b-7807-4920-8333-30fa236e7190.png)

**Comment by sephiroth-j** [comment]
There is also the issue [JENKINS-48512](https://issues.jenkins.io/browse/JENKINS-48512) from 2017 that requests to support build args for `DockerBuilderPublisher`.

**Comment by activedecay** [comment]
> This feature does not work or works in such a way that it is useless in most cases. When you specify references to job parameters in the parameter value, e.g: `KEY=${VALUE}`, the Docker build command passes not `KEY=REAL_VALUE` but just the string `KEY=${VALUE}`, which makes no sense.

I agree. It should use environment variables. Setting this field to `PACKAGES=${PACKAGES}` executes 

    RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y $PACKAGES

and prints the error

    E: Couldn't find any package by regex '${PACKAGES}'

So, literally, `${PACKAGES}`, instead of expanding it.
